Added C++ version of getBestRMS()

psavery opened the pull request:
getBestRMS() was previously a Python-only function. This
adds a C++ version of the function.

<!--
Thanks for contributing a pull request! 
-->
#### Reference Issue
<!-- Example: Fixes #1234 -->


#### What does this implement/fix? Explain your changes.


#### Any other comments?

psavery commented:
Here is the Python function that this is a C++ version of: https://github.com/rdkit/rdkit/blob/870fc26a9b7cb5b04fdcd2f85acd9e2847dca608/rdkit/Chem/AllChem.py#L93

Also, a part of this function was used for a guide as well: https://github.com/rdkit/rdkit/blob/b76af93513ab372f32ada6bcb451a2881fbddeb2/Code/GraphMol/Wrap/Mol.cpp#L114

greglandrum commented on Code/GraphMol/MolAlign/AlignMolecules.cpp:
```diff
@@ -69,6 +69,51 @@ double alignMol(ROMol &prbMol, const ROMol &refMol, int prbCid, int refCid,
+    bool recursionPossible = true;
+    bool useChirality = false;
+    bool useQueryQueryMatches = false;
+    int maxMatches = 1000;
```
The 1000 here and the 1e6 below don't make sense together.
Should pick one and I think it makes sense to add it as an optional argument to the function.

greglandrum commented on Code/GraphMol/MolAlign/AlignMolecules.cpp:
```diff
@@ -69,6 +69,51 @@ double alignMol(ROMol &prbMol, const ROMol &refMol, int prbCid, int refCid,
 
+double getBestRMS(ROMol& refMol, ROMol& probeMol,
+                  int refId, int probeId,
+                  const std::vector<MatchVectType>& _matches)
```
please remove the underscore from the argument name

greglandrum commented on Code/GraphMol/MolAlign/AlignMolecules.h:
```diff
@@ -100,6 +100,33 @@ double alignMol(ROMol &prbMol, const ROMol &refMol, int prbCid = -1,
+  <b>Returns</b>
+  Best RMSD value found
+*/
+double getBestRMS(ROMol& refMol, ROMol& probeMol,
```
as above: please remove the underscore from the parameter name.

psavery commented:
I made an update here. A few questions about it:

Are you okay with me passing the std::vector for matches by copy instead of by reference? I was originally passing it with an underscore so I could immediately make a copy afterwards that I could edit (in case we need to run a SubstructMatch()). But if I just pass it by copy to begin with (like I have it set up right now in the update), I don't have to worry about that.

I added maxMatches as an optional argument, but it is only important if the user doesn't specify the "matches" argument themselves. About the large maxMatches warning: even if the user were to pass in something like 1e10 for maxMatches, I think it still makes sense to warn them if SubstructMatch found a lot of them (like more than 1e6) because it will be time-consuming. However, I'm fine with removing that check (matches.size() > 1e6) if you don't think it is necessary. In theory, matches.size() should never be greater than maxMatches because you give maxMatches to SubstructMatch, right?

I was looking into adding a test for getBestRMS() to https://github.com/rdkit/rdkit/blob/master/Code/GraphMol/MolAlign/testMolAlign.cpp

But the test actually fails when I run it without my newly added test. In fact, after building, if I run "make test", I get a lot of failures. Perhaps that is not the way you run the tests? Or I am doing something wrong?

greglandrum commented:
I think passing `matches` by const reference is nicer. There's something that feels "ugly" to me about the "copy and use the argument as a temporary variable in the function" approach.

Agreed about warning if there are a huge number of matches; even if this is less than the max possible number.

I'll do a separate comment about tests.

greglandrum commented:
Getting the tests to run as part of your build environment requires that you have RDBASE set properly and that your LD_LIBRARY_PATH (or equivalent) is set such that the tests that need share libraries find the right versions.

If you're running the python tests too, you'll need to have PYTHONPATH set as well.

psavery commented:
Okay. I pushed another update. I changed the argument name of "matches" to be "maps", and I make a copy of it inside the function called "matches" that can be used to do a substruct search if maps is empty.

I also started adding a test for getBestRMS(). Creating the environment variables as you suggested fixed the problems I was having with the tests.

However, I'm getting some strange results from the test that I am trying to understand and will keep looking at. I should be getting 0.657861 for the RMS, but I keep getting ~4.8 instead.

It looks like the substruct search is working properly. I compared it with the Python version of getBestRMS(), and they both find the same matches (as far as I can tell). They both find 4 MatchVectTypes and they are all identical (as far as I can tell, once again).

But when I try to run alignMol() with the atom mappings from the substruct search, I get around 4.8 for the RMS values for the 4 mappings in the C++ version (doesn't happen in the Python version - the Python version gets around 0.66 for each of them).

In fact, when I just run the C++ alignMol() without any mappings, I get the correct RMS (~0.65). So something strange is going on - and I need to keep looking into it. I will let you know if I figure out what it is.

psavery commented:
Ah, I figured it out.

alignMol() takes the probe mol first and the reference mol second, but SubstructMatch takes the reference mol first and the probe mol second. I was using the same ordering for both.

I suppose this is a great example of what testing is good for :-).

Will push an update soon.

psavery commented:
Alright Greg. I updated the function, and it seems to work fine now. I also added a test which I think is decent - I found an example where AlignMol() returns a different RMSD than GetBestRMS(), and in the test, I make sure that the GetBestRMS() value is returned.

I also added a Python version and a python test. The Python test seems to work fine on my end.

Although I seemed to get it working fine, I am not all that familiar with Boost::python, so please let me know if I should have done something differently (especially the _translateAtomMapVector(python::object atomMapVec) function).

Note: I think users who would call GetBestRMS() in Python will still be able to call it with the same imports because AllChem.py does "from rdkit.Chem.rdMolAlign import *". So hopefully people won't be having import problems with the new function placement.

One other difference, though, is that this GetBestRMS() has an additional argument: the maxMatches argument that comes immediately before maps. This could potentially break some people's code if their code calls GetBestRMS() with all of the arguments. The reason I put maxMatches before maps, though, is that maxMatches is only useful if a user does NOT define the maps argument (or if they pass in an empty vector/list).

Anyways, let me know if you think there should be more changes.

psavery commented:
And sorry - I meant this to be one commit. I will squash them tomorrow, but the changes will be the same.

greglandrum commented:
Don't worry about squashing. Github does that automatically when merging the PR.

greglandrum commented on Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp:
```diff
@@ -197,6 +208,20 @@ double AlignMolecule(ROMol &prbMol, const ROMol &refMol, int prbCid = -1,
   return rmsd;
 }
 
+double GetBestRMS(ROMol &prbMol, ROMol &refMol, int prbId = -1,
```
There's no need to provide default values for the arguments here. The only place this function will be called is from the boost wrapper, and you provide defaults in your `def` call. 
There's nothing wrong with doing it, and other places in the code do this too, but it is redundant.

greglandrum commented on Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp:
```diff
@@ -197,6 +208,20 @@ double AlignMolecule(ROMol &prbMol, const ROMol &refMol, int prbCid = -1,
+double GetBestRMS(ROMol &prbMol, ROMol &refMol, int prbId = -1,
+                  int refId = -1, int maxMatches = 1000000,
+                  python::object map = python::list()) {
+  std::vector<MatchVectType> aMapVec = _translateAtomMapVector(map);
```
The default value of map should be `python::object()` (this corresponds to `None` in Python), there's a reminder about this below.
Then here you would do:
```
std::vector<MatchVectType> aMapVec;
if(map != python::object()) aMapVec = _translateAtomMapVector(map);
```

greglandrum commented on Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp:
```diff
@@ -637,6 +662,39 @@ BOOST_PYTHON_MODULE(rdMolAlign) {
+      "GetBestRMS", RDKit::GetBestRMS,
+      (python::arg("prbMol"), python::arg("refMol"), python::arg("prbId") = -1,
+       python::arg("refId") = -1, python::arg("maxMatches") = 1000000,
+       python::arg("map") = python::list()), docString.c_str());
```
default value of `map` should be `python::object()`

greglandrum commented:
This is almost there. Thanks Patrick!

On the topic of argument ordering: I would prefer to have the extra argument at the end of the argument list within the Python API, just to avoid breaking code. It's not a big deal because you're technically supposed to supply argument names when you provide keyword arguments, but people don't always do this. I will accept the PR either way, so it's your call.

psavery commented:
Okay! I updated it with the suggestions you made. For the sake of consistency, I moved maxMatches to be the last argument in the C++ getBestRMS() also.

I also added a check at the end of getBestRMS() so it can skip re-aligning the molecule to the same alignment twice if the last alignment was the best one.

Let me know if there are any more changes needed.

greglandrum commented:
Looks great. Thanks @psavery !

psavery commented:
Thanks Greg! Hopefully I'll get to make more contributions in the future as well!

gedeck commented:
A while ago, I started a rewrite of the molecule alignment code. It's pretty far, however not quite finished. If someone wants to have a look at it, here is the code:

https://github.com/gedeck/rdkit/tree/gedeck-torsionFingerprints-merged/Code

https://github.com/gedeck/rdkit/tree/gedeck-torsionFingerprints-merged/Code/GraphMol/MolAlign

It's quite behind the current master and will most certainly have conflicts with this PR.
